src/libcore/slice/mod.rs

```diff
@@ -2663,6 +2663,60 @@ impl<T> [T] {
     {
         self.iter().is_sorted_by_key(f)
     }
+
+    /// Returns the index of the partition point according to the given predicate
+    /// (the index of the first element of the second partition).
+    ///
+    /// The slice is assumed to be partitioned according to the given predicate.
+    /// This means that all elements for which the predicate returns true are at the start of the slice
+    /// and all elements for which the predicate returns false are at the end.
+    /// For example, [7, 15, 3, 5, 4, 12, 6] is a partitioned under the predicate x % 2 != 0
+    /// (all odd numbers are at the start, all even at the end).
+    ///
+    /// If this slice is not partitioned, the returned result is unspecified and meaningless,
```

Comment:
> /// If this slice is not partitioned, the returned result is unspecified and meaningless,
> /// as this method performs a kind of binary search.

I think we'd better change to "If this slice is not ordered ..." as we actually want to tell the end-user an ordered slice is required for this method.

Reply:
The comment is right -- the slice has to be partitioned, and not ordered. Raison d'être of `partition_point` is that it works with non-`Ord` things. To make this more precise, we might link to https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.is_partitioned.

Reply:
Indeed in 95% cases `partition_point` is used to get `lower/upper_bound`. However in 5% cases it looks more partition point than lower bound.
For example, we can get all valid values from `[3, 1, 4, 1, 5, -1, -1, -1, -1, -1]`.
Actually this array is sorted by `|&x| x != -1` in descending order (true is larger than false), but many user would think that the array is partitioned at index 5 by whether -1 or not.

What do you think about adding like this?
> Sorted slice in ascending order is always partitioned by whether less than any value, and in descending order is always partitioned by whether greater than any value. For example, `[1, 2, 4, 8, 16]` is partitioned by `|&x| x < -10`, `|&x| x < 2`, `|&x| x < 5`, and `|&x| x < -100` and other.

Reply:
Thanks for the kind explanation and sorry for my misunderstanding. Yes, the slice requires partitioned instead of sorted. 

However, I still confused by the comment `If this slice is not partitioned, the returned result is unspecified and meaningless`. If we did call this method on a not partitioned slice, the result should be an `Option<size>` rather than an unspecified or meaningless value.

Reply:
Binary search runs in O(log n) **by assuming** ordered. If it checks whether is sorted, it must slower than O(n).